### PR TITLE
python3Packages.python-lsp-server: make some dependencies optional

### PR DESCRIPTION
**Pull request by figsoda**
###### Motivation for this change

Some dependencies of python-lsp-server can be optional
https://github.com/python-lsp/python-lsp-server/blob/v1.2.1/setup.py#L47

`cq-editor` fails to build with `nixpkgs-review` because it is [already broken](https://hydra.nixos.org/eval/1698305?filter=cq-editor)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/python-lsp-server/default.nix [hunk 1]
```diff
@@ -22,8 +22,21 @@
...
+, withPyflakes ? true
+, withPylint ? true
+, withRope ? true
+, withYapf ? true
 }:
 
+let
+  inherit (lib) optional;
+in
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-lsp-server/default.nix:39]:
```suggestion
```
just use lib.optional

File: pkgs/development/python-modules/python-lsp-server/default.nix [hunk 2]
```diff
@@ -22,8 +22,21 @@
...
 , yapf
+, withAutopep8 ? true
+, withFlake8 ? true
+, withMccabe ? true
+, withPycodestyle ? true
+, withPydocstyle ? true
+, withPyflakes ? true
+, withPylint ? true
+, withRope ? true
+, withYapf ? true
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-lsp-server/default.nix:33]: Those are a lot of options. I doubt anyone will test any combination of them. Why do we want them at all? Is there a good reason you want to disable one?

Comment (figsoda) [reply pkgs/development/python-modules/python-lsp-server/default.nix:33]: [python-lsp-black](https://github.com/python-lsp/python-lsp-black#usage) recommends users to not have `autopep8` and `yapf` installed, there could be other plugins that doesn't work well with the some of these dependencies. The optional dependencies might also add extra lint messages that users might dislike (I don't care about my scripts' complexity, but mccabe is included by default and there is not easy way of removing it).

Comment (SuperSandro2000) [reply pkgs/development/python-modules/python-lsp-server/default.nix:33]: Could we group the options into less logical ones?

Comment (figsoda) [reply pkgs/development/python-modules/python-lsp-server/default.nix:33]: Not sure if there's a good way to do it

File: pkgs/development/python-modules/python-lsp-server/default.nix [hunk 3]
```diff
@@ -62,6 +70,21 @@ buildPythonPackage rec {
     pytestCheckHook
   ];
 
+  disabledTests =
+    lib.optional (!withPycodestyle) "test_workspace_loads_pycodestyle_config";
+
+  disabledTestPaths =
+    lib.optional (!withAutopep8) "test/plugins/test_autopep8_format.py"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-lsp-server/default.nix:77]:
```suggestion
  disabledTests = lib.optional (!withPycodestyle) "test_workspace_loads_pycodestyle_config";

  disabledTestPaths = lib.optional (!withAutopep8) "test/plugins/test_autopep8_format.py"
```

File: pkgs/development/python-modules/python-lsp-server/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/python-lsp-server/default.nix:33]: hmmm... Then we just do it like this until we have a better idea.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 135276` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>cq-editor: log was empty</li>
  </ul>
</details>
<details>
<summary>
13 packages built:
</summary>
<ul>
<li>python38Packages.pyls-flake8</li>
<li>python38Packages.pyls-isort</li>
<li>python38Packages.pyls-spyder</li>
<li>python38Packages.pylsp-mypy</li>
<li>python38Packages.python-lsp-black</li>
<li>python38Packages.python-lsp-server</li>
<li>python38Packages.spyder</li>
<li>python39Packages.pyls-flake8</li>
<li>python39Packages.pyls-isort</li>
<li>python39Packages.pyls-spyder</li>
<li>python39Packages.pylsp-mypy</li>
<li>python39Packages.python-lsp-black</li>
<li>python39Packages.python-lsp-server</li>
</ul>
</details>

**Pull request merged**